### PR TITLE
migrate the build command to use the build daemon

### PR DESCRIPTION
**Pull request by jakemac53**

### CODE REVIEW

File: webdev/lib/src/command/build_command.dart [hunk 1]
```diff
@@ -98,71 +33,76 @@ class BuildCommand extends Command<int> {
...
+          ..hoist = configuration.outputInput.isNotEmpty);
+      }
+      client.registerBuildTarget(DefaultBuildTarget((b) => b
+        ..target = configuration.outputInput
+        ..outputLocation = outputLocation?.toBuilder()));
+      client.startBuild();
+      var exitCode = 0;
+      var gotBuildStart = false;
+      await for (final result in client.buildResults) {
+        var targetResult = result.results.firstWhere(
```
Comment (jakemac53) [inline webdev/lib/src/command/build_command.dart]: I am not sure if this is the correct thing to do here - it seems really weird to get multiple build results from a single build result event?

File: webdev/lib/src/command/build_command.dart [hunk 2]
```diff
@@ -98,71 +33,76 @@ class BuildCommand extends Command<int> {
...
-      stderr.writeln('Please file an issue with reproduction steps at '
-          'https://github.com/dart-lang/build/issues\n\n');
-      final error = e[0];
-      final trace = e[1] as String;
-      stderr.writeln(error);
-      stderr.writeln(Trace.parse(trace).terse);
-      if (exitCode == 0) exitCode = 1;
-    });
+    final arguments =
+        buildRunnerArgs(pubspecLock, configuration, includeOutput: false);
```
Comment (grouma) [inline webdev/lib/src/command/build_command.dart]: I think you can remove the `includeOutput` option altogether now.

File: webdev/lib/src/command/build_command.dart [hunk 3]
```diff
@@ -98,71 +33,76 @@ class BuildCommand extends Command<int> {
...
+        },
+      );
+      OutputLocation outputLocation;
+      if (configuration.outputPath != null) {
+        outputLocation = OutputLocation((b) => b
+          ..output = configuration.outputPath
+          ..useSymlinks = false
+          ..hoist = configuration.outputInput.isNotEmpty);
+      }
+      client.registerBuildTarget(DefaultBuildTarget((b) => b
```
Comment (grouma) [inline webdev/lib/src/command/build_command.dart]: Can you not have multiple targets? e.g. `webdev build -o web:foo -o test:bar`?

File: webdev/lib/src/command/build_command.dart (hunk 1 above)
Comment (grouma) [reply webdev/lib/src/command/build_command.dart]: A client can register for multiple targets. Internally each target can either pass or fail. Thus the result of a single build, which happens in parallel, will be collected into this abstraction.

File: webdev/lib/src/command/build_command.dart [hunk 4]
```diff
@@ -98,71 +33,76 @@ class BuildCommand extends Command<int> {
...
+        ..target = configuration.outputInput
+        ..outputLocation = outputLocation?.toBuilder()));
+      client.startBuild();
+      var exitCode = 0;
+      var gotBuildStart = false;
+      await for (final result in client.buildResults) {
+        var targetResult = result.results.firstWhere(
+            (buildResult) => buildResult.target == configuration.outputInput,
+            orElse: () => null);
+        if (targetResult == null) continue;
```
Comment (grouma) [inline webdev/lib/src/command/build_command.dart]: This code reads odd to me. Can you not just do a firstWhere the result contains the desired target and the status isn't started?

File: webdev/lib/src/serve/dev_workflow.dart [hunk 5]
```diff
@@ -32,7 +29,7 @@ Future<BuildDaemonClient> _startBuildDaemon(
         logHandler(recordLevel, trimLevel(recordLevel, serverLog.log));
       },
     );
-  } on OptionsSkew {
+  } on OptionsSkew catch (_) {
```
Comment (grouma) [inline webdev/lib/src/serve/dev_workflow.dart]: Why is this change necessary?

**Comment by jakemac53** [comment]
> I'm surprised there are no test changes. They all work as expected?

I was just letting travis find the errors and ill fix them up - but there don't seem to be a lot, mostly logging related.

File: webdev/lib/src/command/build_command.dart (hunk 3 above)
Comment (jakemac53) [reply webdev/lib/src/command/build_command.dart]: Hmm you should be able to but I don't really understand how that works today. I will take a closer look.

File: webdev/lib/src/command/build_command.dart (hunk 1 above)
Comment (jakemac53) [reply webdev/lib/src/command/build_command.dart]: oh ok that makes sense - will I only get results that contain my target then? So I can remove this check and the null check?

Comment (jakemac53) [reply webdev/lib/src/command/build_command.dart]: Will need some additional handling for building multiple targets too once I add that in.

File: webdev/lib/src/command/build_command.dart (hunk 4 above)
Comment (jakemac53) [reply webdev/lib/src/command/build_command.dart]: I wanted to make sure I don't pick up a result for an ongoing build that I didn't start (it could be out of date).

File: webdev/lib/src/serve/dev_workflow.dart (hunk 5 above)
Comment (jakemac53) [reply webdev/lib/src/serve/dev_workflow.dart]: I guess it isn't just feels weird lol. I will remove (I was trying to actually log the error to see if I got better info but it didn't help).

File: webdev/lib/src/command/build_command.dart (hunk 3 above)
Comment (jakemac53) [reply webdev/lib/src/command/build_command.dart]: Ah, webdev only supports a single output dir today per build, so that simplifies things

File: webdev/lib/src/command/build_command.dart (hunk 1 above)
Comment (jakemac53) [reply webdev/lib/src/command/build_command.dart]: Going to just leave this as is actually, it should be plenty safe although maybe a bit defensive

File: webdev/lib/src/command/build_command.dart (hunk 4 above)
Comment (jakemac53) [reply webdev/lib/src/command/build_command.dart]: Added a comment and restructured this so its a bit nicer

Comment (grouma) [reply webdev/lib/src/command/build_command.dart]: Looks better. Thanks!

File: webdev/test/e2e_test.dart [hunk 6]
```diff
@@ -52,14 +52,14 @@ void main() {
...
-            '[SEVERE] Unable to create merged directory at ${d.sandbox}.'));
+            contains('Unable to create merged directory at ${d.sandbox}.')));
     await expectLater(
         process.stdout,
         emitsThrough(
             'Choose a different directory or delete the contents of that '
             'directory.'));
 
-    await process.shouldExit(73);
+    await process.shouldExit(isNot(0));
```
Comment (grouma) [inline webdev/test/e2e_test.dart]: Is this technically a breaking change?

Comment (natebosch) [reply webdev/test/e2e_test.dart]: This was potentially valuable, though I don't know if there were any actual users. Is it possible to still support it?

File: webdev/lib/src/logging.dart [hunk 7]
```diff
@@ -46,3 +47,24 @@ void _colorLog(Level level, String message, {bool verbose}) {
...
+
+/// Detects if the [ServerLog] contains a [Level] and returns the
+/// resulting value.
+///
+/// If the [ServerLog] does not contain a [Level], null will be returned.
+Level levelForLog(ServerLog serverLog) {
+  var log = serverLog.log;
+  Level recordLevel;
+  for (var level in Level.LEVELS) {
+    if (log.startsWith('[$level]')) {
```
Comment (natebosch) [inline webdev/lib/src/logging.dart]: We're going to end up building this string over and over. I wonder if we want something like:

```dart
final levelsByLabel = Map.fromIterable(Level.LEVELS, key: (l) => '[$l]'); // should be static

...
var messageLabel = levelsByLabel.keys.firstWhere(log.startsWith);
return messageLabel == null ? null : levelsByLabel[messageLabel];
```

Then we can look forward to writing it instead as 

```dart
var levelsByLabel = {for (var l in Level.LEVELS) '[$l]': l};

**Comment by kevmoo** [comment]
Do we want to get this in for v2?

File: webdev/lib/src/logging.dart (hunk 7 above)
Comment (jakemac53) [reply webdev/lib/src/logging.dart]: I just moved these utilities around in this pr since they aren't specific to the serve command any more, we could follow up with something but I think we have a lot of other much slower things going on in the logging than this :D.

Also probably a switch statement or something would probably be faster than a map for that if we are concerned about perf?

File: webdev/test/e2e_test.dart (hunk 6 above)
Comment (jakemac53) [reply webdev/test/e2e_test.dart]: We can't reasonably support it as it stands right now I don't think, we don't get exit codes from the build daemon as far as I could tell.

**Comment by jakemac53** [comment]
> Do we want to get this in for v2?

yes
